runtime/hsa-runtime/core/driver/kfd/amd_kfd_driver.cpp

```diff
@@ -42,40 +42,120 @@
 
 #include "core/inc/amd_kfd_driver.h"
 
-#include <sys/ioctl.h>
-
 #include <memory>
 #include <string>
 
+#include <link.h>
+#include <sys/ioctl.h>
+
 #include "hsakmt/hsakmt.h"
 
-#include "core/inc/amd_cpu_agent.h"
-#include "core/inc/amd_gpu_agent.h"
 #include "core/inc/amd_memory_region.h"
-#include "core/inc/exceptions.h"
 #include "core/inc/runtime.h"
 
+extern r_debug _amdgpu_r_debug;
+
 namespace rocr {
 namespace AMD {
 
 KfdDriver::KfdDriver(std::string devnode_name)
     : core::Driver(core::DriverType::KFD, devnode_name) {}
 
-hsa_status_t KfdDriver::Init() { return HSA_STATUS_SUCCESS; }
+hsa_status_t KfdDriver::Init() {
+  HSAKMT_STATUS ret = hsaKmtRuntimeEnable(
+      &_amdgpu_r_debug, core::Runtime::runtime_singleton_->flag().debug());
+
+  if (ret != HSAKMT_STATUS_SUCCESS && ret != HSAKMT_STATUS_NOT_SUPPORTED)
+    return HSA_STATUS_ERROR;
+
+  uint32_t caps_mask = 0;
+  hsaKmtGetRuntimeCapabilities(&caps_mask);
+  core::Runtime::runtime_singleton_->KfdVersion(
+      ret != HSAKMT_STATUS_NOT_SUPPORTED,
+      !!(caps_mask & HSA_RUNTIME_ENABLE_CAPS_SUPPORTS_CORE_DUMP_MASK));
+
+  if (hsaKmtGetVersion(&version_) != HSAKMT_STATUS_SUCCESS)
+    return HSA_STATUS_ERROR;
+
+  if (version_.KernelInterfaceMajorVersion == kfd_version_major_min &&
+      version_.KernelInterfaceMinorVersion < kfd_version_major_min)
+    return HSA_STATUS_ERROR;
+
+  core::Runtime::runtime_singleton_->KfdVersion(version_);
+
+  if (version_.KernelInterfaceMajorVersion == 1 &&
+      version_.KernelInterfaceMinorVersion == 0)
+    core::g_use_interrupt_wait = false;
+
+  bool xnack_mode = BindXnackMode();
+  core::Runtime::runtime_singleton_->XnackEnabled(xnack_mode);
 
-hsa_status_t KfdDriver::DiscoverDriver() {
-  if (hsaKmtOpenKFD() == HSAKMT_STATUS_SUCCESS) {
-    std::unique_ptr<Driver> kfd_drv(new KfdDriver("/dev/kfd"));
-    core::Runtime::runtime_singleton_->RegisterDriver(kfd_drv);
+  return HSA_STATUS_SUCCESS;
+}
+
+hsa_status_t KfdDriver::ShutDown() {
+  HSAKMT_STATUS ret = hsaKmtRuntimeDisable();
+  if (ret != HSAKMT_STATUS_SUCCESS)
+    return HSA_STATUS_ERROR;
+
+  ret = hsaKmtReleaseSystemProperties();
+
+  if (ret != HSAKMT_STATUS_SUCCESS)
+    return HSA_STATUS_ERROR;
+
+  return Close();
+}
+
+hsa_status_t KfdDriver::DiscoverDriver(std::unique_ptr<core::Driver> &driver) {
+  driver = std::unique_ptr<core::Driver>(new KfdDriver("/dev/kfd"));
+
+  if (driver->Open() == HSA_STATUS_SUCCESS) {
     return HSA_STATUS_SUCCESS;
   }
+
+  driver.reset(nullptr);
```

Comment:
More of stylistic opinion, but this would be idiomatic C++ code:
```cpp
  auto driver_tmp = std::unique_ptr<core::Driver>(new KfdDriver("/dev/kfd"));
  if (driver_tmp ->Open() != HSA_STATUS_SUCCESS) {
    return HSA_STATUS_ERROR;
  }
  driver = std::move(driver_tmp);
```

Reply:
Fixed